### ISSUE TITLE
Refactor current implementation of tag infrastructure and storage of Ad Blocking Recovery tags to use `Setting` instead of option fields.

### ISSUE THREAD

**Issue by kuasha420** [google/site-kit-wp#7156]
## Feature Description

Currently (or more accurately, once #7134 is merged), Ad Blocking Recovery tag and Error Protection tag is being encoded and stored in separate option fields are we added two dedicated getters to retrieve and decode the tags for use in the `AdSense` module class. However, as pointed out by @aaemnnosttv, there is no benefit to storing them separately and we should refactor to use a dedicated `Setting` class to store them together as a serialized value. We can hand off the encoding/decoding business to the class as well which will make things more clear and reduce repetition and duplicate code.

Serializing them will also mitigate the problem of tempering and length. 

Lastly, we should also refactor the usage of the methods introduced in #6902 that will be used in #6963, but it should be drop in replacement without much issues. 

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The Ad Blocking recovery tag storage should be refactored to use a dedicated `Setting` : `Google\Site_Kit\Modules\AdSense\Ad_Blocking_Recovery_Tag`
	* The option name should be `googlesitekit_adsense_ad_blocking_recovery_tag`
* The value should be stored as an array with the following schema
	* `tag : string`
	* `error_protection_code : string`
* The setting values should be stored as base64 encoded strings, although the setting itself should be stored as any other serialized array by WP
* The `get` and `set` methods should always accept and return unencoded values
* The AdSense module should no longer expose these values by dedicated public methods ~~but by a single method to get the `Ad_Blocking_Recovery_Tag` instance, similar to `get_settings`~~

## Implementation Brief

* Create `Google\Site_Kit\Modules\AdSense\Ad_Blocking_Recovery_Tag` class in `includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php` location.
    * It should extend `Google\Site_Kit\Core\Storage` class.
    * `OPTION` should be `googlesitekit_adsense_ad_blocking_recovery_tag`.
    * Return `object` from `get_type` method.
    * Return empty an array with `tag` and `error_protection_code` key and values of both being an empty string from `get_default` method. 
    * The `get_sanitize_callback` should ensure that both `$option['tag']` and `$option['error_protection_code']` is set and of string type. 
    * Override the `set` method:
        * Ensure both `$option['tag']` and `$option['error_protection_code']` are of type string. 
        * Encode them with `base64_encode` and then call `parent::set` with the updated `$option` array. 
    * Override the `get` method:
        * Get the currently saved `$option` with `parent::get()`.
        * If the return value is empty array, return this as is to indicate that the value is not set. 
        * Otherwise, decode both `$option['tag']` and `$option['error_protection_code']` using `base64_decode`, passing `true` for a strict decode that fails on invalid character and length.
        * If both `tag` and `error_protection_code` were decoded successfully (non empty string), update the `$options` with the decoded strings and return the array. 
        * If there one or both of the decoding failed, return an empty array. 
* In `Google\Site_Kit\Modules\AdSense` class:
    * Add protected instance variable `$ad_blocking_recovery_tag` which is of `Ad_Blocking_Recovery_Tag` type.
    * Create a protected `get_ad_blocking_recovery_tag` method:
        * This should work similarly to how `get_settings` work, namely:
        * If `$this->ad_blocking_recovery_tag` not an instance of `Ad_Blocking_Recovery_Tag`, it should be set as a new instance of `Ad_Blocking_Recovery_Tag` passing `$this->options` while constructing it. 
        * Return `$this->ad_blocking_recovery_tag`. 
    * In the `register` method:
        * Register the `Ad_Blocking_Recovery_Tag` using `$this->get_ad_blocking_recovery_tag()->register()` method. 
    * In The `POST:sync-ad-blocking-recovery-tags` route of `parse_data_response` method:
        * Replace the current usage of `$this-options->set()` with the newly added `$this->get_ad_blocking_recovery_tag()->set()`, passing both `tag` and `error_protection_code` together as an array. 
    * Remove the following pieces of codes as they are no longer used. 
        * `AD_BLOCKING_RECOVERY_TAG_OPTION` constant
        * `AD_BLOCKING_RECOVERY_ERROR_PROTECTION_TAG_OPTION` constant 
        * `get_ad_blocking_recovery_tag` method
        * `get_ad_blocking_recovery_error_protection_tag` method

### Test Coverage

* Add test coverage for `Ad_Blocking_Recovery_Tag`.
* Update any failing tests 
* Remove tests for the removed methods. 

## QA Brief

* Complete AdSense setup.
* Sync the ABD tags on the server using the following console command, 
    * `await googlesitekit.data.dispatch('modules/adsense').syncAdBlockingRecoveryTags()`
* Ensure the response is `{ "success": true }`.
* Check the `googlesitekit_adsense_ad_blocking_recovery_tag` options from the DATABASE. it should be populated with an encrypted string. 

## Changelog entry

* N/A

**Comment by aaemnnosttv** [comment]
> * Return empty array from `get_default` method.

@kuasha420 I think we want to return an array with the same schema (similar to module settings). For the default they'd just be empty strings.

We also want to make sure the `Ad_Blocking_Recovery_Tag` class is registered (since it is a `Setting`), which can probably happen in the `AdSense::register` since it's only relevant when the module is active (unlike regular settings which are always registered).

**Comment by kuasha420** [comment]
@aaemnnosttv Nice catch, I had intended to add the `register` call in the `AdSense::register` but must've forgotten it. 

As for the default value, I don't have any issue with your reasoning, as it's more consistent. Although just returning an empty array would also work as we're not exposing them client side like the module settings. 

IB Updated. :white_check_mark:

**Comment by kuasha420** [comment]
Noting for any engineers QA'ng this, the requirement about encoding/decoding in AC was changed to encryption/decryption in CR phase and was greenlit after consensus from everyone. See [here](https://github.com/google/site-kit-wp/pull/7175#discussion_r1234252556) and [here](https://10up.slack.com/archives/CBKKQEBR9/p1687249800262429) for the context. Cheers.

**Comment by wpdarren** [comment]
### QA Update:  ⚠️ 

@kuasha420 While the ticket is set to `QA:Eng`, the QAB steps seemed easy enough so thought I'd have a go.  

I ran the console code on the AdSense settings (and also on the dashboard) and an error occurs. 

![image](https://github.com/google/site-kit-wp/assets/73545194/e28dacac-b820-45b3-8c66-8f12e39c612e)

I also ran the code provided in another ticket to set the tag placement, but that didn't seem to help. 

The user who it says does not have permission, does, because it's the same user.  

That said, I used the tester plugin to set the account and site to 'ready' 

Any ideas why this is erroring out - or is it something that can only be tested on a local site? 🤔

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@kuasha420 as per our chat on Slack.  I realise you're unable to recreate this but I do on a few different InstaWP sites. 

Would you be able to re-check the QAB, and also see if you can recreate on an InstaWP site rather than local.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@kuasha420 okay, I am getting closer to getting this into approved, but still an issue. 

In the database I do not see an option for  `googlesitekit_adsense_ad_blocking_recovery_tag` this is when looking in the wp-options table.  I also searched for the option above and could not find anything.   I did see some settings in adsense settings but not this particular option.  Any ideas? 

![image](https://github.com/google/site-kit-wp/assets/73545194/d92ec54f-73d6-4843-835c-422e3eee0d9b)

**Verified:**

Okay, after a conversation with @kuasha420 it materialises that you cannot use the tester plugin to test this. 😁 so, I have a site that has an account and site set to ready and when using this I was able to see the success mesage. 

`Ensure the response is { "success": true }.`

![image](https://github.com/google/site-kit-wp/assets/73545194/ef0395e1-a4b6-4d19-a27e-a0a533d4bdbd)

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@kuasha420 because the QAB does not specifically say what I should see in the database, please could I check that this is correct? I don't want to assume. 

> Check the `googlesitekit_adsense_ad_blocking_recovery_tag` options from the DATABASE. it should be populated with an encrypted string.

![image](https://github.com/google/site-kit-wp/assets/73545194/90d9fbe6-465b-41a0-8e4f-49c7348acc48)

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

I have a site that has an account and site set to ready and when using this I was able to see the success message.

`Ensure the response is { "success": true }.`

![image](https://github.com/google/site-kit-wp/assets/73545194/528902e7-ccf6-4d50-82dc-ce64246773fa)

- The googlesitekit_adsense_ad_blocking_recovery_tag options from the DATABASE. it should be populated with an encrypted string.

![image](https://github.com/google/site-kit-wp/assets/73545194/56cb799b-0c70-490d-bf60-f85c28e194e0)
